### PR TITLE
Fuse Nemotron latent MoE projection and shared add

### PR DESCRIPTION
**Pull request by samnordmann**
## Motivation

Nemotron Ultra projects routed MoE output from latent width 2048 back to hidden width 8192, then adds the full-width shared-expert output. This currently launches a GEMM followed by a separate BF16 add in every latent MoE layer:

```python
# Before
routed = F.linear(routed_latent, weight)
output = routed + shared

# After
output = torch.addmm(shared, routed_latent, weight.T, out=shared)
```

The second form uses the GEMM beta input and writes back to the consumed shared-expert buffer, removing one `[tokens, 8192]` memory pass and kernel launch.

## Modifications

- Add an unquantized BF16 linear path that accumulates into a caller-owned addend when the active runtime route is cuBLAS.
- Use it for Nemotron-H's latent output projection and shared-expert add.
- Restrict the in-place path to NVIDIA CUDA and only when the current optimized BF16 dispatcher declines Split-K, Hopper GEMV, and CuTeDSL.
- Preserve normal module forwarding for LoRA wrappers, quantized projections, compilation, AITER, bias, autograd, non-BF16 inputs, and incompatible layouts.
- Add focused wrapper, dispatcher-fallback, platform-scope, and changing-input CUDA graph tests.

## Accuracy Tests

- Updated GB200 focused suite at [`c79a33d4bd`](https://github.com/samnordmann/sglang/commit/c79a33d4bdde05b3d11f575ccfe1fc4d3fd44453): `10 passed`, including changing-input CUDA graph replay, seven fallback classes, and a LoRA-wrapper regression.
- Six paired serving runs matched input/output token counts and completed without request errors. Generated text matched for 418/480 decode requests and 119/120 prefill requests.

The fused operation is numerically equivalent within BF16 tolerance but not bit-identical: the previous path rounds the GEMM output to BF16 before the add, while `addmm` performs the addition in the GEMM accumulator and rounds once. Model-level quality evaluation is therefore still required before merge.

The benchmarked NVFP4 checkpoint does exercise the unquantized path: its [quantization config](https://huggingface.co/nvidia/NVIDIA-Nemotron-3-Ultra-550B-A55B-NVFP4/blob/main/hf_quant_config.json) lists all 48 `fc2_latent_proj` layers in `exclude_modules` and none in `quantized_layers`; SGLang therefore selects [`UnquantizedLinearMethod`](https://github.com/sgl-project/sglang/blob/c79a33d4bdde05b3d11f575ccfe1fc4d3fd44453/python/sglang/srt/layers/quantization/modelopt_quant.py#L974-L996) for those projections.

### Single op
Latency improvement:

| Tokens | Median improvement | Process min–max |
|---:|---:|---:|
| 16 | **+0.82%** | `+0.43% to +24.88%` |
| 32 | **+18.82%** | `+7.70% to +19.86%` |
| 256 | **+16.63%** | `+16.62% to +16.98%` |

### E2E

NVIDIA Nemotron 3 Ultra 550B A55B NVFP4, one 4xGB200 node, TP4/EP4, FlashInfer attention, `flashinfer_trtllm` MoE, EAGLE 5, and CUDA graphs. Same-node alternating process pairs:

| Workload ISL/OSL, concurrency | Metric | Baseline | Candidate | Mean paired improvement (95% CI) |
|---|---|---:|---:|---:|
| 8192/2048, c16 | Output throughput | 2317.54 tok/s | 2343.91 tok/s | **+1.14%** |
| 8192/2048, c16 | TPOT | 5.811 ms | 5.746 ms | **-1.11%** |
|  50000/32, c4 | Input throughput | 25264.08 tok/s | 25469.31 tok/s | **+0.81%**|
| 50000/32, c4 | Mean TTFT | 5666.73 ms | 5619.02 ms | **-0.84%** |

## Checklist

- [x] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [x] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [x] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations). No documentation change is required for this model-local implementation detail.
- [ ] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed). Speed and focused correctness results are included; model-level quality evaluation remains pending.
- [x] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

## Review and Merge Process

1. Ping Merge Oncalls to start the process. See the [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process).
2. Get approvals from [CODEOWNERS](https://github.com/sgl-project/sglang/blob/main/.github/CODEOWNERS) and other reviewers.
3. Trigger CI tests with [comments](https://docs.sglang.io/developer_guide/contribution_guide.html#how-to-trigger-ci-tests) or contact authorized users to do so.
4. After green CI and required approvals, ask Merge Oncalls or people with Write permission to merge the PR.

---
### CI States

Latest PR Test (Base): :white_check_mark: [Run #34041683530](https://github.com/sgl-project/sglang/actions/runs/34041683530)
Latest PR Test (Extra): :white_check_mark: [Run #34041686031](https://github.com/sgl-project/sglang/actions/runs/34041686031)
Latest PR Test (AMD ROCm 7.2): :x: [Run #34041683734](https://github.com/sgl-project/sglang/actions/runs/34041683734)

### CODE REVIEW

File: test/registered/unit/models/test_nemotron_h_shared_add.py [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+        def forward(self, x):
+            return 2 * x + 1, None
+
+    routed = torch.randn(4, 8)
+    shared = torch.randn_like(routed)
+    moe = SimpleNamespace(fc2_latent_proj=LoraLikeProjection())
+
+    output = NemotronHMoE._apply_latent_projection(moe, routed, shared)
+
+    torch.testing.assert_close(output, 2 * routed + 1 + shared, rtol=0, atol=0)
```
Comment (nvpohanh) [inline test/registered/unit/models/test_nemotron_h_shared_add.py]: [by Codex] **Severity: functional | Confidence: High**

This registered pytest file has no `if __name__ == "__main__":` entry. `collect_tests(..., sanity_check=True)` rejects it before scheduling, so Base CI cannot run any of these tests. Add `import sys` and `sys.exit(pytest.main([__file__, "-v"]))` at the end.

**Review by nvpohanh — COMMENTED** [review]
[by Codex] One inline review finding.

**Comment by nvpohanh** [comment]
@samnordmann could you address the comment?

**Comment by nvpohanh** [comment]
/rerun-test test/registered/unit/models/test_nemotron_h_shared_add.py

File: python/sglang/srt/layers/quantization/unquant.py [hunk 2]
```diff
@@ -403,6 +403,67 @@ def apply(
...
+            or addend.shape != (x.shape[0], layer.weight.shape[0])
+            or bias is not None
+            or x.requires_grad
+            or addend.requires_grad
+            or layer.weight.requires_grad
+        ):
+            return False
+
+        backend = get_bf16_gemm_backend()
+        if backend in (Bf16GemmBackend.AUTO, Bf16GemmBackend.TORCH):
```
Comment (mmangkad) [inline python/sglang/srt/layers/quantization/unquant.py]: TORCH is what deterministic inference selects, and AUTO lands here too on non-SM100. The addmm below goes through `aten::addmm.out`, which batch-invariant mode doesn't override.

File: python/sglang/srt/layers/quantization/unquant.py [hunk 3]
```diff
@@ -403,6 +403,67 @@ def apply(
...
+            return True
+        if not backend.is_optimized() or _use_cutedsl_bf16_gemm is None:
+            return False
+
+        m, n, k = x.shape[0], layer.weight.shape[0], layer.weight.shape[1]
+        # Preserve any custom dispatcher route. Accumulate into the addend only
+        # when the optimized backend would fall through to F.linear/cuBLAS.
+        if _enable_bf16_splitk_gemm and use_flashinfer_pr4266_bf16_gemm(m, n, k):
+            return False
+        if _use_hopper_bf16_gemv is not None and _use_hopper_bf16_gemv(m, n, k):
```
Comment (mmangkad) [inline python/sglang/srt/layers/quantization/unquant.py]: `_use_hopper_bf16_gemv` is only ever set when the backend is GEMV, and GEMV is never `is_optimized()`. This gate also duplicates `_bf16_gemm_dispatch_impl`'s routing, and its static half re-runs on every call.

File: python/sglang/srt/layers/quantization/unquant.py [hunk 4]
```diff
@@ -403,6 +403,67 @@ def apply(
...
+        The fast path overwrites ``addend``. Callers must treat it as consumed.
+        Unlike ``apply_into``, this uses the existing buffer as the GEMM beta
+        input (beta=1) instead of treating it as write-only output (beta=0).
+        """
+        if self._can_apply_with_addend(layer, x, addend, bias):
+            # Reuse addend as beta input and output; cuBLAS rounds once after GEMM.
+            return torch.addmm(addend, x, layer.weight.t(), out=addend)
+
+        output = self.apply(layer, x, bias)
+        output.add_(addend)
```
Comment (mmangkad) [inline python/sglang/srt/layers/quantization/unquant.py]: L455 says addend is consumed. This path leaves it untouched, and L138 asserts that.

File: python/sglang/srt/models/nemotron_h.py [hunk 5]
```diff
@@ -331,6 +332,27 @@ def _forward_core_shared_routed_overlap(
 
         return final_hidden_states, shared_output
 
+    def _apply_latent_projection(
+        self,
+        final_hidden_states: torch.Tensor,
+        shared_output: torch.Tensor | None,
+    ) -> torch.Tensor:
+        projection = self.fc2_latent_proj
+        if shared_output is not None and isinstance(projection, ReplicatedLinear):
```
Comment (mmangkad) [inline python/sglang/srt/models/nemotron_h.py]: This skips `ReplicatedLinear.forward`, which is where `skip_bias_add` is handled. LoRA only misses the branch because `ReplicatedLinearWithLoRA` doesn't happen to subclass `ReplicatedLinear`.

File: test/registered/unit/models/test_nemotron_h_shared_add.py [hunk 6]
```diff
@@ -0,0 +1,178 @@
...
+        monkeypatch.setattr(unquant, "_use_cutedsl_bf16_gemm", lambda *args: True)
+        monkeypatch.setattr(
+            unquant,
+            "_cutedsl_bf16_gemm",
+            lambda x, weight, bias: F.linear(x, weight, bias),
+        )
+    elif fallback == "hopper_gemv":
+        monkeypatch.setattr(unquant, "_use_hopper_bf16_gemv", lambda *args: True)
+        linear_output = F.linear(routed, projection.weight)
+        monkeypatch.setattr(method, "apply", lambda layer, x, bias: linear_output)
```
Comment (mmangkad) [inline test/registered/unit/models/test_nemotron_h_shared_add.py]: `aiter`, `hopper_gemv` and `splitk` all stub out `apply`, so none of them reach the dispatch in their name. L164 has the same gap, with a hand written stand-in for `ReplicatedLinearWithLoRA`.

**Review by mmangkad — CHANGES_REQUESTED** [review]
Paired run on 4x GB300 TP4 against the same tree with only the two source files reverted: gsm8k 96.59% to 96.51%, TPOT down 2.4% at c1 and 1.7% at c16, so the speedup holds, but the gate still takes the fused path under `--enable-deterministic-inference`, where `addmm(out=)` skips batch-invariant mode.

File: python/sglang/srt/layers/quantization/unquant.py (hunk 2 above)
Comment (samnordmann) [reply python/sglang/srt/layers/quantization/unquant.py]: Right, I confirmed that `torch.addmm(..., out=...)` takes the `aten::addmm.out` overload, while batch-invariant mode overrides `aten::mm` and the default `aten::addmm`, not the "out" overload.

Now I gate the reuse path wiht `is_batch_invariant_mode_enabled()`, so deterministic path retains the unfused version. I added a regression test for that.

File: python/sglang/srt/layers/quantization/unquant.py (hunk 3 above)
Comment (samnordmann) [reply python/sglang/srt/layers/quantization/unquant.py]: Ok
I removed the duplicated `_can_apply_with_addend` checks and moved `addend` handling into the existing dispatcher, so backend selection is done only once.

File: python/sglang/srt/layers/quantization/unquant.py (hunk 4 above)
Comment (samnordmann) [reply python/sglang/srt/layers/quantization/unquant.py]: Agreed. the comment was misleading. I clarified that only `addmm(out=addend)` overwrites `addend`.

File: python/sglang/srt/models/nemotron_h.py (hunk 5 above)
Comment (samnordmann) [reply python/sglang/srt/models/nemotron_h.py]: Good catch. I moved `addend` handling into `ReplicatedLinear.forward` and limited the optimization to the exact base class, preserving wrapper and subclass behavior. wdyt?

File: test/registered/unit/models/test_nemotron_h_shared_add.py (hunk 6 above)
Comment (samnordmann) [reply test/registered/unit/models/test_nemotron_h_shared_add.py]: I rewrote the tests to exercise the real dispatch logic, mocking only the backend kernel entry points, and verify that each expected backend is called once. The Lora case now uses an active `ReplicatedLinearWithLoRA`.

File: python/sglang/srt/layers/linear.py
Comment (mmangkad) [inline python/sglang/srt/layers/linear.py:1]: Can we not touch this file?

File: python/sglang/srt/layers/quantization/unquant.py [hunk 7]
```diff
@@ -402,6 +414,49 @@ def apply(
...
+            or x.dtype != torch.bfloat16
+            or layer.weight.dtype != torch.bfloat16
+            or addend.dtype != torch.bfloat16
+            or not addend.is_contiguous()
+            or addend.shape != (x.shape[0], layer.weight.shape[0])
+            or bias is not None
+            or x.requires_grad
+            or addend.requires_grad
+            or layer.weight.requires_grad
+        ):
```
Comment (mmangkad) [inline python/sglang/srt/layers/quantization/unquant.py]: We can simplify this

**Review by mmangkad — APPROVED** [review]
